src/statement.ts

```diff
@@ -0,0 +1,240 @@
+  private readonly parameters?: Parameters;
+
+  constructor(
+    config: Config & Partial<AutoWaitConfig>,
+    sql: string,
+    parameters?: Parameters
+  ) {
+    this.config = config;
+    this.sql = sql;
+    this.parameters = parameters;
+  }
+
+  /**
+   * Bind a set of values to the parameters of the prepared statement.
+   * We follow the SQLite convention for prepared statements parameter binding.
+   * We support Ordered (?NNNN), Anonymous (?), and Named (@name, :name, $name) parameters.
+   * @param values A variadic list of values to bind. May include base types, and objects.
+   * @returns A bound Statement.
+   */
+  bind<T = S>(...values: ValuesType[]): Statement<T> {
+    const parameters = getParameters(...values);
+    return new Statement(this.config, this.sql, parameters);
+  }
+
+  /**
+   * Resolve a bound statement to a SQL string.
+   * @returns A valid SQL string.
+   */
+  toString(): string {
+    try {
+      return bindValues(this.sql, this.parameters);
+    } catch (cause) {
+      throw new Error("BIND_ERROR", { cause });
+    }
+  }
+
+  toObject(): { sql: string; parameters?: Parameters } {
+    return {
+      sql: this.sql,
+      parameters: this.parameters,
+    };
+  }
+
+  async #parseAndExtract(): Promise<ExtractedStatement> {
+    const statementWithBindings = this.toString();
+    const { type, statements, tables } = await normalize(statementWithBindings);
+    // Stick with original if a create statement, otherwise, use the parsed version
+    // This is because the parser injects keywords that are not spec compliant
+    // See https://github.com/tablelandnetwork/go-sqlparser/issues/41
+    const sql =
+      type === "create" ? statementWithBindings : statements.join(";");
+    return { type, sql, tables };
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async all<T = S, K extends keyof T = keyof T>(
+    colName?: undefined,
+    opts?: Signal
+  ): Promise<Result<T>>;
+  async all<T = S, K extends keyof T = keyof T>(
+    colName: K,
+    opts?: Signal
+  ): Promise<Result<T[K]>>;
+  async all<T = S, K extends keyof T = keyof T>(
+    colName?: K,
+    opts: Signal = {}
+  ): Promise<Result<T | T[K]>> {
+    try {
+      const start = performance.now();
+      const { sql, type, tables } = await this.#parseAndExtract();
+      switch (type) {
+        case "read": {
+          const config = await extractReadonly(this.config, {
+            type,
+            tables,
+          });
+          const results = await queryAll<T>(config, sql, opts);
+          if (colName != null) {
+            return wrapResult(
+              extractColumn(results, colName),
+              performance.now() - start
+            );
+          }
+          return wrapResult(results, performance.now() - start);
+        }
+        default: {
+          let receipt = await exec(this.config, { type, sql, tables });
+          if (this.config.autoWait ?? false) {
```

Comment:
Since D1 doesn't have the concept of deferred execution or transactions like we do, the downstream ORMs and things will return right after the smart contract call is made. If you turn on autoWait, they'll actually only resolve once the validator resolves the transaction. This is nice because it makes the API feel like a web2 API. But you can also turn this off and explicitly wait for things yourself, or not wait at all.